### PR TITLE
CBG-919: Integrated Prometheus Exporter

### PR DESCRIPTION
**Pull request by JRascagneres**
Created new struct based stat structure. Ensured there is no additional overhead compared to expvars for add / get / set. Less likely to add data to incorrect group accidentally ( see below issue ). Allows for future changes to support other stat formats

---

We have stat 'StatKeyConflictWriteCount'. Its init'd to 0 for CBL Push and the const is under CBL Push but we only seem to ever use it under DatabaseStats. Have gone with using it in Database stats for now but need to confirm this before any merge.

### CODE REVIEW

**Review by bbrks — COMMENTED** [review]
General approach looks good:
- No additional overhead for Add/Set
- Less error prone linking exported stat names to their actual stats compared to mapping lists
- I like how struct-based stats give you compile time guarantees about what stat belongs to which group, along with the autocomplete hints that brings
- The sgwStat struct type allows us to tack on additional metadata for each stat if we need it for future formats

Changes to make:
- Make stats thread safe via atomic operations (like expvar.Add/Set did under the hood)
- Allow float stats and the ability to add/set float values to float stats
- The other specifics like renames, etc. we've already talked about

File: base/exporter.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	CpuPercentUtil                      *sgwIntStat `json:"cpu_percent_util"`
+}
+
+type DbStats struct {
+}
+
+type sgwIntStat struct {
+	statFQDN      string
+	statDesc      *prometheus.Desc
+	dbName        string
```
Comment (adamcfraser) [inline base/exporter.go]: As discussed, since we also need to support labels like "replication":"replicationName", probably need to change this to a generic 'labels' map.

File: base/exporter.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+			DeltaCacheHit:             NewIntStat("delta_sync", "delta_cache_hit", dbName, prometheus.CounterValue, 0),
+			DeltaCacheMiss:            NewIntStat("delta_sync", "delta_sync_miss", dbName, prometheus.CounterValue, 0),
+			DeltaPushDocCount:         NewIntStat("delta_sync", "delta_push_doc_count", dbName, prometheus.CounterValue, 0),
+		}
+	}
+	return d.DeltaSyncStats
+}
+
+func (d *DbStats) SharedBucketImport() *SharedBucketImportStats {
+	if d.SharedBucketImportStats == nil {
```
Comment (adamcfraser) [inline base/exporter.go]: A simple nil check here isn't concurrency-safe (which is why the existing functionality was using sync.Once).

File: base/exporter.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+		}
+	}
+
+	return d.GsiStats.Stats[queryName]
+}
+
+func (g *GsiStats) MarshalJSON() ([]byte, error) {
+	ret := map[string]interface{}{}
+	for queryName, queryMap := range g.Stats {
+		ret[queryName+"_query_count"] = queryMap.QueryCount
```
Comment (adamcfraser) [inline base/exporter.go]: These names seem like they don't align with the existing names (e.g. StatKeyN1qlQueryCountExpvarFormat and StatKeyViewQueryCountExpvarFormat) - I think the _query_ should be removed?

File: base/exporter.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+			QueryCount:      NewIntStat("gsi_views", queryName+"_query_count", dbName, prometheus.CounterValue, 0),
+			QueryErrorCount: NewIntStat("gsi_views", queryName+"_query_error_count", dbName, prometheus.CounterValue, 0),
+			QueryTime:       NewIntStat("gsi_views", queryName+"_query_time", dbName, prometheus.CounterValue, 0),
+		}
+	}
+
+	return d.GsiStats.Stats[queryName]
+}
+
+func (g *GsiStats) MarshalJSON() ([]byte, error) {
```
Comment (adamcfraser) [inline base/exporter.go]: This approach is a bit clunky, but I think that's mainly because we need to provide backward compatibility to the existing stat format.  Ideally I think we'd probably just use something like this for similar functionality going forward:
```
"gsi_views": {
  "channels": {
     "count":1
     "error_count":0,
     "time":1234,
   },
   "access": {
        ...
    }
}

File: db/query.go [hunk 5]
```diff
@@ -238,11 +238,12 @@ func (context *DatabaseContext) N1QLQueryWithStats(queryName string, statement s
 
 	results, err = gocbBucket.Query(statement, params, consistency, adhoc)
 	if err != nil {
-		context.DbStats.StatsGsiViews().Add(fmt.Sprintf(base.StatKeyN1qlQueryErrorCountExpvarFormat, queryName), 1)
+		context.DbStats.NewStats.GSIStats(queryName).QueryErrorCount.Add(1)
 	}
 
-	context.DbStats.StatsGsiViews().Add(fmt.Sprintf(base.StatKeyN1qlQueryCountExpvarFormat, queryName), 1)
-	context.DbStats.StatsGsiViews().Add(fmt.Sprintf(base.StatKeyN1qlQueryTimeExpvarFormat, queryName), time.Since(startTime).Nanoseconds())
+	context.DbStats.NewStats.GSIStats(queryName).QueryErrorCount.Add(1)
```
Comment (adamcfraser) [inline db/query.go]: This should just be QueryCount.

File: db/query.go [hunk 6]
```diff
@@ -238,11 +238,12 @@ func (context *DatabaseContext) N1QLQueryWithStats(queryName string, statement s
...
+	context.DbStats.NewStats.GSIStats(queryName).QueryTime.Add(int(time.Since(startTime).Nanoseconds()))
```
Comment (adamcfraser) [inline db/query.go]: An advantage of this approach is that we can just call context.DbStats.NewStats.GSIStats(queryName) once in this method (to minimize map lookups), and use that in case there are multiple stat updates.

File: base/exporter.go [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+			ImportProcessingTime: NewIntStat("shared_bucket_import", "import_processing_time", dbName, prometheus.GaugeValue, 0),
+			ImportHighSeq:        NewIntStat("shared_bucket_import", "import_high_seq", dbName, prometheus.CounterValue, 0),
+			ImportPartitions:     NewIntStat("shared_bucket_import", "import_partitions", dbName, prometheus.GaugeValue, 0),
+		}
+	}
+
+	return d.SharedBucketImportStats
+}
+
+func (d *DbStats) GSIStats(queryName string) *QueryStat {
```
Comment (adamcfraser) [inline base/exporter.go]: Should be named QueryStats, since it's also used for view stats.

File: db/query_test.go [hunk 8]
```diff
@@ -55,9 +51,9 @@ func TestQueryChannelsStatsView(t *testing.T) {
 	closeErr := results.Close()
 	assert.NoError(t, closeErr, "Close error")
 
-	channelQueryCountAfter := base.ExpvarVar2Int(db.DbStats.StatsGsiViews().Get(queryCountExpvar))
-	channelQueryTimeAfter := base.ExpvarVar2Int(db.DbStats.StatsGsiViews().Get(queryTimeExpvar))
-	channelQueryErrorCountAfter := base.ExpvarVar2Int(db.DbStats.StatsGsiViews().Get(errorCountExpvar))
+	channelQueryCountAfter := db.DbStats.NewStats.GSIStats(fmt.Sprintf("%s.%s", DesignDocSyncGateway(), ViewChannels)).QueryCount.Value.(int)
```
Comment (adamcfraser) [inline db/query_test.go]: If QueryTypeChannels is used for GSI, it's probably reasonable to create constants or vars for the view equivalents, instead of building them with fmt.Sprintf every time.

File: base/exporter.go (hunk 2 above)
Comment (JRascagneres) [reply base/exporter.go]: Yeah. I need to go back and add various concurrency things such as the talked about atomics etc.

File: base/exporter.go (hunk 3 above)
Comment (JRascagneres) [reply base/exporter.go]: You might be right. I haven't done a careful pass over checking that all the names are the same as they are now. I'll make sure I check over these a bunch before I submit for a full review

File: base/exporter.go (hunk 4 above)
Comment (JRascagneres) [reply base/exporter.go]: Yeah. That suggested snippet you added is basically the structure that the structs I added follow but as you say, added this to ensure its backwards compatible.

File: db/query.go (hunk 5 above)
Comment (JRascagneres) [reply db/query.go]: Good catch. This is something that I'll have to double check too, its easier to check things like this in the PR review so I'll make sure to do that!

File: base/exporter.go (hunk 7 above)
Comment (JRascagneres) [reply base/exporter.go]: Very good point

File: base/exporter.go [hunk 9]
```diff
@@ -0,0 +1,714 @@
+package base
```
Comment (adamcfraser) [inline base/exporter.go]: Since we're defining all stats here (and not just Prometheus), maybe move the general stats functionality to a stats.go, and potentially stick prometheus specific functionality into stats_prometheus.go?

File: base/exporter.go [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+
+type DbStats struct {
+	dbName                  string
+	CacheStats              *CacheStats                   `json:"cache,omitempty"`
+	CBLReplicationPullStats *CBLReplicationPullStats      `json:"cbl_replication_pull,omitempty"`
+	CBLReplicationPushStats *CBLReplicationPushStats      `json:"cbl_replication_push,omitempty"`
+	DatabaseStats           *DatabaseStats                `json:"database,omitempty"`
+	DeltaSyncStats          *DeltaSyncStats               `json:"delta_sync,omitempty"`
+	QueryStats              *QueryStats                   `json:"gsi_views,omitempty"`
+	DbReplicatorStats       map[string]*DbReplicatorStats `json:"db_replicator_stats"`
```
Comment (adamcfraser) [inline base/exporter.go]: I think this should be just `json:"replications"`

File: base/exporter.go [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+			dbName:        dbName,
+			statValueType: statValueType,
+		},
+		Val: initialValue,
+	}
+	prometheus.MustRegister(stat)
+	return stat
+}
+
+func (s *SgwFloatStat) Describe(ch chan<- *prometheus.Desc) {
```
Comment (adamcfraser) [inline base/exporter.go]: There seems to be more in common than different between the SgwFloatStat and SgwIntStat implementations.  Can some or all of the common code be pushed up to SgwStat?

File: base/exporter.go [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+func (s *SgwFloatStat) Collect(ch chan<- prometheus.Metric) {
+	var labelVals []string
+	if s.dbName != "" {
+		labelVals = append(labelVals, s.dbName)
+	}
+	ch <- prometheus.MustNewConstMetric(s.statDesc, s.statValueType, s.Val, labelVals...)
+}
+
+func (s *SgwFloatStat) Set(newV float64) {
+	s.Val = newV
```
Comment (adamcfraser) [inline base/exporter.go]: Locking missing when modifying the float values.

File: base/exporter.go (hunk 1 above)
Comment (adamcfraser) [reply base/exporter.go]: I think this still applies.  It's reasonable to have the constructor just accept dbName for ease of use, but internally it seems more appropriate to store slices of label names/label values.

File: rest/config.go [hunk 13]
```diff
@@ -1041,6 +1041,9 @@ func RunServer(config *ServerConfig) {
 
 	go sc.PostStartup()
 
+	base.Consolef(base.LevelInfo, base.KeyAll, "")
+	go config.Serve(":1234", CreateMetricHandler(sc))
```
Comment (adamcfraser) [inline rest/config.go]: Port should be configurable, and presumably default to something other than 1234.

File: manifest/default.xml [hunk 14]
```diff
@@ -135,5 +135,37 @@
...
+  <project name="client_golang" path="godeps/src/github.com/prometheus/client_golang" revision="master" remote="prometheus" />
+  <project name="client_model" path="godeps/src/github.com/prometheus/client_model" revision="master" remote="prometheus" />
+  <project name="common" path="godeps/src/github.com/prometheus/common" revision="master" remote="prometheus" />
+  <project name="procfs" path="godeps/src/github.com/prometheus/procfs" revision="master" remote="prometheus" />
+  <project name="perks" path="godeps/src/github.com/beorn7/perks" revision="master" remote="beorn7" />
+  <project name="xxhash" path="godeps/src/github.com/cespare/xxhash" revision="master" remote="cespare" />
+  <project name="protobuf" path="godeps/src/github.com/golang/protobuf" revision="master" remote="golang" />
+  <project name="protobuf-go" path="godeps/src/google.golang.org/protobuf" revision="master" remote="protocolbuffers" />
+
+  <!-- Missing:
```
Comment (adamcfraser) [inline manifest/default.xml]: Is this ready for another round of updates?

File: base/exporter.go [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+type ReplicatorStats struct {
+	toplevelMap *expvar.Map
+}
+
+func (rs *ReplicatorStats) MarshalJSON() ([]byte, error) {
+	return []byte(rs.toplevelMap.String()), nil
+}
+
+func (rs *ReplicatorStats) Describe(ch chan<- *prometheus.Desc) {
+	return
```
Comment (adamcfraser) [inline base/exporter.go]: What's the functionality associated with Describe that we're not supporting?

File: base/exporter.go [hunk 16]
```diff
@@ -0,0 +1,714 @@
...
+}
+
+func (rs *ReplicatorStats) Collect(ch chan<- prometheus.Metric) {
+	rs.toplevelMap.Do(func(value expvar.KeyValue) {
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(prometheus.BuildFQName("sgw", "replication", "sgr_docs_checked_sent"), "sgr_docs_checked_sent", []string{"replication"}, nil),
```
Comment (adamcfraser) [inline base/exporter.go]: Why are these generated at collection time, unlike the rest of the stats?  I realize they aren't getting published to prometheus at a per-stat level, but is there a way to only generate the prometheus desc once?

File: base/exporter.go [hunk 17]
```diff
@@ -0,0 +1,714 @@
...
+	DeltasSent                *SgwIntStat `json:"deltas_sent"`
+}
+
+type QueryStats struct {
+	Stats map[string]*QueryStat
+	mutex sync.Mutex
+}
+
+type DbReplicatorStats struct {
+	NumAttachmentBytesPushed *SgwIntStat `json:"num_attachment_bytes_pushed"`
```
Comment (adamcfraser) [inline base/exporter.go]: Why the change in naming of these stats?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
A few comments on some of the details.  It looks like a few things are still WIP/pending cleanup - have a look and we can plan an additional review.

File: manifest/default.xml (hunk 14 above)
Comment (JRascagneres) [reply manifest/default.xml]: Yeah, forgot to switch to couchbase deps

File: base/exporter.go (hunk 15 above)
Comment (JRascagneres) [reply base/exporter.go]: Its a good question and the docs are rather sparse on the purpose of this method, however, I'm pretty confident that we don't lose any functionality. Its more about internal management.

File: base/exporter.go (hunk 17 above)
Comment (JRascagneres) [reply base/exporter.go]: I guess you're questioning why they're named differently in json rather than the exporter? If so there's no real reason. I'll change them to be the same.

File: base/exporter.go (hunk 11 above)
Comment (JRascagneres) [reply base/exporter.go]: Some of the constructor can be moved out

File: base/exporter.go (hunk 16 above)
Comment (JRascagneres) [reply base/exporter.go]: The work here is done on collect time as this is still based on an expvar map. This means we need to loop over it all at collection time as we don't know what stats are available prior to this. I have however moved the description generation to when the ReplicatorStats are initialized so that they are only generated once

File: db/query_test.go (hunk 8 above)
Comment (JRascagneres) [reply db/query_test.go]: I've added a const to use for the formatting bit: "%s.%s" not sure its nice to have the full thing as a const incase names change.

File: base/stats.go [hunk 18]
```diff
@@ -0,0 +1,747 @@
...
+	GlobalStats     GlobalStat          `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication"`
+
+	initDBStats               sync.Once
+	registeredReplicatorStats sync.Once
+}
+
+func (s *SgwStats) String() string {
+	bytes, _ := JSONMarshal(s)
```
Comment (adamcfraser) [inline base/stats.go]: I think this needs to be JSONMarshalCanonical, to provide the key ordering that's expected for expvar maps.

Comment (adamcfraser) [inline base/stats.go]: Why is it safe to ignore the error here?  Seems like something that should log info at minimum.

File: base/stats.go [hunk 19]
```diff
@@ -0,0 +1,747 @@
...
+const (
+	StatViewFormat = "%s.%s"
+)
+
+type SgwStats struct {
+	GlobalStats     GlobalStat          `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication"`
+
+	initDBStats               sync.Once
```
Comment (adamcfraser) [inline base/stats.go]: In general it's preferable to avoid sync.Once unless absolutely needed.  In this case it seems like we should just have a constructor for SgwStats that initializes DbStats and ReplicatorStats, and maybe stick omitempty on ReplicatorStats?

File: base/stats.go [hunk 20]
```diff
@@ -0,0 +1,747 @@
...
+		}
+	}
+}
+
+func (d *DbStats) Cache() *CacheStats {
+	return d.CacheStats
+}
+
+func (d *DbStats) initCBLReplicationPullStats() {
+	if d.CBLReplicationPullStats == nil {
```
Comment (adamcfraser) [inline base/stats.go]: What is the nil check doing for us here?  If we know initCBLReplicationPullStats is only going to be called once (at initialization time), it's not needed.  If we don't know that, the nil check alone isn't sufficient to avoid races.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: base/expvar.go [hunk 21]
```diff
@@ -239,49 +61,169 @@ func init() {
...
-	stats.Set(StatKeyGoMemstatsHeapInUse, ExpvarIntVal(0))
-	stats.Set(StatKeyGoMemstatsHeapReleased, ExpvarIntVal(0))
-	stats.Set(StatKeyGoMemstatsStackInUse, ExpvarIntVal(0))
-	stats.Set(StatKeyGoMemstatsStackSys, ExpvarIntVal(0))
-	stats.Set(StatKeyGoMemstatsPauseTotalNs, ExpvarIntVal(0))
-	stats.Set(StatKeyErrorCount, ExpvarIntVal(0))
-	stats.Set(StatKeyWarnCount, ExpvarIntVal(0))
+
+	SyncGatewayStats.GlobalStats.ResourceUtilization.AdminNetworkInterfaceBytesReceived = NewIntStat(
+		"resource_utilization",
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/expvar.go]: Nope. Will add that. Also will shift this resource utils into stats.go to be consistent with where the rest of it is

File: base/stats.go [hunk 22]
```diff
@@ -0,0 +1,747 @@
...
+func (wrapper *ExpVarMapWrapper) MarshalJSON() ([]byte, error) {
+	return []byte(wrapper.String()), nil
+}
+
+type ReplicatorStats struct {
+	checkedSentDesc,
+	numAttachmentBytesTransferred,
+	numAttachmentsTransferred,
+	numDocsFailedToPush,
+	numDocsPushed *prometheus.Desc
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go]: I have switched to having desc for each item. I think this approach works well as they're all in one place and as we have to initialize the map in that method too it makes sense to also initialize the descriptions there. If you think otherwise I can think again and move them to individual vars.

File: base/stats.go [hunk 23]
```diff
@@ -0,0 +1,747 @@
...
+
+	toplevelMap *expvar.Map
+}
+
+func (rs *ReplicatorStats) MarshalJSON() ([]byte, error) {
+	return []byte(rs.toplevelMap.String()), nil
+}
+
+func (rs *ReplicatorStats) Describe(ch chan<- *prometheus.Desc) {
+	return
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go:268]: https://godoc.org/github.com/prometheus/client_golang/prometheus#Desc
https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector

From these two above essentially it looks like Describe has protection against having duplicate descriptions which isn't allowed as obviously this would cause problems when settings a single stat to the same value. A description essentially is the "key" for the stat.

There is no functionality lost by not having it.

In our implementation using this wouldn't work as we do technically have the same description used twice but when we use it we set different label values. The documentation around Collector seems to suggest that this is fine and we don't have to use it provided we can make guarantees ourselves that there won't be conflicts.

File: base/stats.go [hunk 24]
```diff
@@ -0,0 +1,747 @@
...
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
+}
+
+type ExpVarMapWrapper struct {
+	*expvar.Map
+}
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go:253]: I don't think it can be simplified to that as we have to be able to run a method on the underlying type. Also couldn't come up with a succinct name for the wrapper that would explain its functionality and so I added a comment.

```
// This wrapper ensures that an expvar.Map type can be marshalled into JSON. The expvar.Map has no method to go direct to
// JSON however the map.String() method returns JSON and so we can use this.
```

File: base/stats.go [hunk 25]
```diff
@@ -0,0 +1,747 @@
...
+	SequenceReleasedCount   *SgwIntStat `json:"sequence_released_count"`
+	SequenceReservedCount   *SgwIntStat `json:"sequence_reserved_count"`
+	WarnChannelsPerDocCount *SgwIntStat `json:"warn_channels_per_doc_count"`
+	WarnGrantsPerDocCount   *SgwIntStat `json:"warn_grants_per_doc_count"`
+	WarnXattrSizeCount      *SgwIntStat `json:"warn_xattr_size_count"`
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
```
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go:246]: These two are omitted as is the current case with the external exporter. Comment added.

File: base/stats.go (hunk 18 above)
Comment (JRascagneres) [reply base/stats.go]: Added an error log and returning null json

File: base/stats.go (hunk 20 above)
Comment (JRascagneres) [reply base/stats.go]: This was left over from lazy init. You're right, we can remove these now that we know when they're being initialized

File: base/stats.go (hunk 19 above)
Comment (JRascagneres) [reply base/stats.go]: Makes sense for dbstats. Not sure about replicatorStats. I feel like paying for the sync may be better than initializing the descriptions and map when it might not be needed?

File: base/exporter.go (hunk 15 above)
Comment (JRascagneres) [reply base/exporter.go]: Followed this up in a later comment.

File: base/stats.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (adamcfraser) [reply base/stats.go:253]: You can define a type based on another type and use it as a receiver - you don't need to define as a struct.  There are several examples in the SG codebase I think - ConflictResolverTrype is a simple example, LogPriorityQueue is another.

File: base/stats.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (adamcfraser) [reply base/stats.go]: There are two good reasons to remove the static descriptions from the struct:
- these descriptions are shared across replications, so don't need to be redefined.  They are also only used in one place, so I don't think having them 'all in one place' really adds that much value.
- removing the descriptions simplifies ReplicatorStats to just `type ReplicatorStats *expvar.Map`, and avoids the need to target ReplicatorStats.topLevelMap everywhere

File: base/stats.go (hunk 19 above)
Comment (adamcfraser) [reply base/stats.go]: Initializing the descriptions doesn't seem particularly expensive, and ReplicatorStats is just a map, I think initializing an empty map (with omitempty on the json annotation) is worth it for the simplified code.

File: base/stats.go [hunk 26]
```diff
@@ -0,0 +1,791 @@
...
+	SubsystemSecurity           = "security"
+	SubsystemSharedBucketImport = "shared_bucket_import"
+)
+
+type SgwStats struct {
+	GlobalStats     *GlobalStat         `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication,omitempty"`
+
+	mutex                     sync.Mutex
```
Comment (adamcfraser) [inline base/stats.go]: Is this mutex being used?

File: base/stats.go [hunk 27]
```diff
@@ -0,0 +1,791 @@
...
+		GlobalStats: &GlobalStat{},
+		DbStats:     map[string]*DbStats{},
+	}
+
+	sgwStats.GlobalStats.initResourceUtilizationStats()
+
+	return &sgwStats
+}
+
+func (s *SgwStats) String() string {
```
Comment (adamcfraser) [inline base/stats.go:70]: I think it's worth mentioning in a comment that String() is here to satisfy the expvar.Var interface.

File: base/stats.go [hunk 28]
```diff
@@ -0,0 +1,791 @@
...
+	bytes, err := JSONMarshalCanonical(s)
+	if err != nil {
+		Errorf("Unable to Marshal SgwStats: %v", err)
+		return "null"
```
Comment (adamcfraser) [inline base/stats.go:76]: Is null the correct response in an error scenario (as opposed to empty string)?  Looking at other examples (like expvar.String), it seems like an empty string is returned in an error scenario.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: base/stats.go (hunk 28 above)
Comment (JRascagneres) [reply base/stats.go:76]: expvar.String returns an empty string as its meant to represent a string, however, as this is a struct / object it seems more appropriate to do null.

File: base/stats.go (hunk 26 above)
Comment (JRascagneres) [reply base/stats.go]: No. I was questioning whether I needed one for the DbStats map but came to the conclusion that I don't think its needed.

File: base/stats.go (hunk 19 above)
Comment (JRascagneres) [reply base/stats.go]: Fair enough

File: base/stats.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go:253]: Yes but we need to call String() on the underlying Map and to do this you have to use a struct.

File: base/stats.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (JRascagneres) [reply base/stats.go]: Yeah thats fair. I've made them vars instead and initialize them in the NewSyncGatewayStats() constructor.

File: base/stats.go [hunk 29]
```diff
@@ -0,0 +1,799 @@
...
+	sgwStats.GlobalStats.initResourceUtilizationStats()
+	sgwStats.initReplicationStats()
+
+	return &sgwStats
+}
+
+// This String() is to satisfy the expvar.Var interface which is used to produce the expvar endpoint output.
+func (s *SgwStats) String() string {
+	s.dbStatsMapMutex.Lock()
+	bytes, err := JSONMarshalCanonical(s)
```
Comment (adamcfraser) [inline base/stats.go:72]: JsonMarshalCanonical seems like it's going to be quite a bit more expensive than the expvar.Map.String() we were using previously - in particular because expvar.Map sorts keys as they are added to the map, where this would be sorting the keys each time String() is called.  I'm also curious about memory allocation associated with json Marshal instead of expvar.Map.String().  If you haven't already, can you write a benchmark and generate some profiling to evaluate the performance difference?  Once we see the difference, we can decide whether it's a tolerable degradation for customers doing frequent  _expvar polling as part of their runtime diagnostics today.

File: base/stats.go [hunk 30]
```diff
@@ -0,0 +1,799 @@
...
+		Stats: map[string]*QueryStat{},
+	}
+	d.QueryStats.mutex.Lock()
+	for _, queryName := range queryNames {
+		d.initQueryStat(useViews, queryName)
+	}
+	d.QueryStats.mutex.Unlock()
+}
+
+func (d *DbStats) initQueryStat(useViews bool, queryName string) {
```
Comment (adamcfraser) [inline base/stats.go]: Switch to _initQueryStat to follow the convention used when functions expect the caller to be holding an associated lock.

File: base/stats.go [hunk 31]
```diff
@@ -0,0 +1,799 @@
...
+	if _, ok := d.QueryStats.Stats[queryName]; !ok {
+		dbName := d.dbName
+
+		// Prometheus isn't happy with '.'s in the name and the '.'s come from the design doc version. Design doc isn't
+		// reported to prometheus. Only the view name.
```
Comment (adamcfraser) [inline base/stats.go:757]: Is this what the current exporter is doing?  It feels like a loss of information if we can't differentiate between view and index stats.  I would have thought that replacing `.` with something like `__` down in newSGWStat would be a preferred approach, but if the current exporter is already doing this we could leave as-is for backward compatibility.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: base/stats.go (hunk 31 above)
Comment (JRascagneres) [reply base/stats.go:757]: Yeah, currently there is no differentiation in the exporter about whether a particular gsi_view stat is for, well, gsi and views. Infact the exporter only currently exports _count stats.

File: base/stats.go (hunk 29 above)
Comment (JRascagneres) [reply base/stats.go:72]: Done some benchmarks and have got some profiles ran and it would seem that for some reason the new implementation is actually better? We can chat about this in sync up.

File: base/stats.go [hunk 32]
```diff
@@ -408,43 +410,37 @@ func (s *SgwIntStat) Describe(ch chan<- *prometheus.Desc) {
...
-	defer s.mutex.Unlock()
-	s.Val += newV
+	atomic.AddInt64(&s.Val, newV)
 }
 
 func (s *SgwIntStat) MarshalJSON() ([]byte, error) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	return []byte(fmt.Sprintf("%v", s.Val)), nil
+	return []byte(fmt.Sprintf("%v", atomic.LoadInt64(&s.Val))), nil
```
Comment (bbrks) [inline base/stats.go]: Using FormatInt is quicker and zero alloc, vs. Sprintf:

https://medium.com/swlh/bad-go-frivolous-sprintf-2ad28fedf1a0

It's also what expvar.Int uses.

File: base/stats.go [hunk 33]
```diff
@@ -408,43 +410,37 @@ func (s *SgwIntStat) Describe(ch chan<- *prometheus.Desc) {
...
 }
 
 func (s *SgwIntStat) String() string {
 	return fmt.Sprintf("%v", s.Val)
```
Comment (bbrks) [inline base/stats.go]: Same comment here as the one in `MarshalJSON`

File: base/stats.go [hunk 34]
```diff
@@ -468,43 +464,47 @@ func (s *SgwFloatStat) Describe(ch chan<- *prometheus.Desc) {
...
 func (s *SgwFloatStat) MarshalJSON() ([]byte, error) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	return []byte(fmt.Sprintf("%v", s.Val)), nil
+	return []byte(fmt.Sprintf("%v", math.Float64frombits(atomic.LoadUint64(&s.Val)))), nil
 }
 
 func (s *SgwFloatStat) String() string {
 	return fmt.Sprintf("%v", s.Val)
 }
```
Comment (bbrks) [inline base/stats.go]: Use FormatFloat for these for similar performance reasons. Need to also consider what format the float is represented in, and whether that's valid in the JSON number format.

I would suggest copying the expvar.Float implementations.

**Review by bbrks — DISMISSED** [review]
Some comments for yesterdays commit for the updated atomic based stats.

File: rest/config.go [hunk 35]
```diff
@@ -98,6 +99,7 @@ type ServerConfig struct {
 	Unsupported                *UnsupportedServerConfig `json:"unsupported,omitempty"`            // Config for unsupported features
 	ReplicatorCompression      *int                     `json:"replicator_compression,omitempty"` // BLIP data compression level (0-9)
 	BcryptCost                 int                      `json:"bcrypt_cost,omitempty"`            // bcrypt cost to use for password hashes - Default: bcrypt.DefaultCost
+	MetricsInterface           *string                  `json:"metrics_interface,omitempty"`      // Interface to bind metrics to, default ":4986"
```
Comment (JRascagneres) [inline rest/config.go]: Change json to metricsInterface to align with existing interface options.

**Review by adamcfraser — APPROVED** [review]
